src/server/pfs/server/driver.go

```diff
@@ -1822,78 +1858,84 @@ func (d *driver) putFile(ctx context.Context, file *pfs.File, delimiter pfs.Deli
+		buffer := &bytes.Buffer{}
+		var datumsWritten int64
+		var bytesWritten int64
+		var filesPut int
+		EOF := false
+		var eg errgroup.Group
+		decoder := json.NewDecoder(reader)
+		bufioR := bufio.NewReader(reader)
+
+		indexToRecord := make(map[int]*pfs.PutFileRecord)
+		var mu sync.Mutex
+		for !EOF {
+			var err error
+			var value []byte
+			switch delimiter {
+			case pfs.Delimiter_JSON:
+				var jsonValue json.RawMessage
+				err = decoder.Decode(&jsonValue)
+				value = jsonValue
+			case pfs.Delimiter_LINE:
+				value, err = bufioR.ReadBytes('\n')
+			default:
+				return fmt.Errorf("unrecognized delimiter %s", delimiter.String())
 			}
-		}
-		buffer.Write(value)
-		bytesWritten += int64(len(value))
-		datumsWritten++
-		if buffer.Len() != 0 &&
-			((targetFileBytes != 0 && bytesWritten >= targetFileBytes) ||
-				(targetFileDatums != 0 && datumsWritten >= targetFileDatums) ||
-				(targetFileBytes == 0 && targetFileDatums == 0) ||
-				EOF) {
-			_buffer := buffer
-			index := filesPut
-			eg.Go(func() error {
-				object, size, err := d.pachClient.PutObject(_buffer)
-				if err != nil {
+			if err != nil {
+				if err == io.EOF {
+					EOF = true
+				} else {
 					return err
 				}
-				mu.Lock()
-				defer mu.Unlock()
-				indexToRecord[index] = &pfs.PutFileRecord{
-					SizeBytes:  size,
-					ObjectHash: object.Hash,
-				}
-				return nil
-			})
-			datumsWritten = 0
-			bytesWritten = 0
-			buffer = &bytes.Buffer{}
-			filesPut++
+			}
+			buffer.Write(value)
+			bytesWritten += int64(len(value))
+			datumsWritten++
+			if buffer.Len() != 0 &&
+				((targetFileBytes != 0 && bytesWritten >= targetFileBytes) ||
+					(targetFileDatums != 0 && datumsWritten >= targetFileDatums) ||
+					(targetFileBytes == 0 && targetFileDatums == 0) ||
+					EOF) {
+				_buffer := buffer
+				index := filesPut
+				eg.Go(func() error {
+					object, size, err := d.pachClient.PutObject(_buffer)
+					if err != nil {
+						return err
+					}
+					mu.Lock()
+					defer mu.Unlock()
+					indexToRecord[index] = &pfs.PutFileRecord{
+						SizeBytes:  size,
+						ObjectHash: object.Hash,
+					}
+					return nil
+				})
+				datumsWritten = 0
+				bytesWritten = 0
+				buffer = &bytes.Buffer{}
+				filesPut++
+			}
+		}
+		if err := eg.Wait(); err != nil {
+			return err
 		}
-	}
-	if err := eg.Wait(); err != nil {
-		return err
-	}
 
-	records.Split = true
-	for i := 0; i < len(indexToRecord); i++ {
-		records.Records = append(records.Records, indexToRecord[i])
+		records.Split = true
+		for i := 0; i < len(indexToRecord); i++ {
+			records.Records = append(records.Records, indexToRecord[i])
+		}
 	}
 
+	if oneOff {
+		// oneOff puts only work on branches, so we know branch != "". We pass
+		// a commit with no ID, that ID will be filled in with the head of
+		// branch (if it exists).
+		_, err := d.makeCommit(ctx, "", client.NewCommit(file.Commit.Repo.Name, ""), branch, nil, nil, []string{file.Path}, []*pfs.PutFileRecords{records}, "")
```

Comment:
This is a very minor suggestion, but I think you could even leave the implementation of makeCommit the same if you put the code that creates the hashtree here:
```
			openTree := parentTree.Open()
			for i, record := range records {
				if err := d.applyWrite(recordFiles[i], record, openTree); err != nil {
					return err
				}
			}
			tree, err = openTree.Finish()
			if err != nil {
				return err
			}
			data, err := hashtree.Serialize(tree)
			if err != nil {
				return err
			}
			treeRef, _, err = d.pachClient.WithCtx(ctx).PutObject(bytes.NewReader(data))
			makeCommit(..., treeRef, ...)
```

Reply:
As discussed offline, this used to be how this code worked, but records need to be applied in the same txn that reads and rewrites the branch pointer so that if it races with another `put-file` it doesn't break.